### ISSUE TITLE
test_wsgi_scenario: child segfaults in `objc_initializeAfterForkError` on MacOS

### ISSUE THREAD

**Issue by jgehrcke** [jgehrcke/gipc#52]
Seen on MacOS
- with `platform darwin -- Python 2.7.15, pytest-3.8.1` 
- with `gevent-1.3.6-cp27-cp27m-macosx_10_9_x86_64.whl`

Notably, this is MacOS **system python** (as provided by Travis CI osx builder, September 2018)

`python --version` output:
```
Python 2.7.15

Darwin Traviss-Mac-6.local 17.4.0 Darwin Kernel Version 17.4.0: Sun Dec 17 09:19:54 PST 2017; root:xnu-4570.41.2~1/RELEASE_X86_64 x86_64
```

```
=================================== FAILURES ===================================
____________________ TestComplexUseCases.test_wsgi_scenario ____________________
self = <test_gipc.TestComplexUseCases object at 0x110bdc350>
    def test_wsgi_scenario(self):
        from gevent.pywsgi import WSGIServer
    
        def serve(http_server):
            http_server.serve_forever()
    
        def hello_world(environ, start_response):
            # Generate response in child process.
            with pipe() as (reader, writer):
                start_response('200 OK', [('Content-Type', 'text/html')])
                rg = start_process(
                    target=complchild_test_wsgi_scenario_respgen,
                    args=(writer, ))
                response = reader.get()
                rg.join()
                assert rg.exitcode == 0
            return [response]
    
        http_server = WSGIServer(('localhost', 0), hello_world)
        servelet = gevent.spawn(serve, http_server)
        # Wait for server being bound to socket.
        while True:
            if http_server.address[1] != 0:
                break
            gevent.sleep(0.05)
        client = start_process(
            target=complchild_test_wsgi_scenario_client,
            args=(http_server.address, ))
        client.join()
>       assert client.exitcode == 0
E       assert -6 == 0
E        +  where -6 = <_GProcess(_GProcess-33, stopped[SIGABRT])>.exitcode
test/test_gipc.py:1187: AssertionError
----------------------------- Captured stderr call -----------------------------
12:53:14,150.8  [15344]start_process# Invoke target `<function complchild_test_wsgi_scenario_client at 0x11068f938>` in child process.
12:53:14,152.7  [15344]start# SIGCHLD watcher for 15383 started.
12:53:14,161.6  [15383]_child# _child start. target: `<function complchild_test_wsgi_scenario_client at 0x11068f938>`
12:53:14,164.3  [15383]_child# Delete current hub's threadpool.
12:53:14,164.9  [15383]_child# Destroy hub and default loop.
12:53:14,165.7  [15383]_child# Created new hub and default event loop.
objc[15383]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
objc[15383]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
12:53:14,179.9  [15344]_on_sigchld# SIGCHLD watcher callback for 15383 invoked. Exitcode stored: -6
------------------------------ Captured log call -------------------------------
gipc.py                    275 DEBUG    Invoke target `<function complchild_test_wsgi_scenario_client at 0x11068f938>` in child process.
gipc.py                    453 DEBUG    SIGCHLD watcher for 15383 started.
gipc.py                    468 DEBUG    SIGCHLD watcher callback for 15383 invoked. Exitcode stored: -6
===================== 1 failed, 77 passed in 3.30 seconds ======================
```

**Comment by jgehrcke** [comment]
Highly valuable input from https://bugs.python.org/issue30385#msg293958:

```
Unfortunately, his is a long standing problem when using Python applications that fork on macOS and that end up calling certain system frameworks that under the covers use the system libdispatch which is not fork safe.  In this case, it is most likely due to a call to s urllib function that results in a call to Python's helper module _scproxy to query the macOS system configuration for network proxies.  If your application and system are not using a proxy, the simplest workaround is to disable network proxy lookups by defining the environment variable 'no_proxy' with value '*' in the Python process, for example:

env no_proxy='*' python3.6 ...

Otherwise, restructuring your code to ensure network initialization occurs in the main thread before any forking *might* also prevent the segfault.
```
(Thanks Ned Deily!)

**Comment by jgehrcke** [comment]
https://github.com/ansible/ansible/issues/32499

> This is apparently due to some new security changes made in High Sierra that are breaking lots of Python things that use fork(). Rumor has it that adding export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES before your Ansible run should clear it up. The code that's causing issues is well below Ansible in the stack.

**Comment by jgehrcke** [comment]
> Further comment: this is almost certainly a bug in the python which apple distributes. Apple has changed their fork implementation to disallow forking with active threads. Taking the latest python distributed from python.org seems to fix this error without needing to set the environment variable (tested on python 3.6).

Ha.

**Comment by jgehrcke** [comment]
> The underlying root cause is Mac OS X's libdispatch.dylib. A ton of system APIs (like this proxy one, or GUI libraries, etc), use it. It seems the proxy settings API use it to manage inter-process communication. libdispatch has code that explicitly "poisons" the process if it forks. I think this is because it internally spawns threads, so the forked child state is unreliable, and they figure it is better for it to crash than to fail randomly. This is the classic "don't mix threads and fork" issue, its just that the threads are hidden inside a bunch of system APIs.

(https://bugs.python.org/issue27126)
Love the entire thread there!

**Comment by jgehrcke** [comment]
> BTW. Using fork without exec is unsafe in any program using threads, people tend to run into this more on OSX because Apple's libraries use background threads while other Unix-y platforms don't tend to do that.

**Comment by jgehrcke** [comment]
Fixed via #55 and #58.
